### PR TITLE
feat(cohere-transcribe): CoreML export + host-side pipeline fix + q8 hybrid

### PR DESCRIPTION
**Pull request by Alex-Wengg**
## Summary

CoreML conversion pipeline for Cohere Transcribe 03-2026 (14-language encoder-decoder ASR), plus the host-side preprocessing and decode fixes required for the shipped weights to actually produce correct output.

Scope is intentionally limited to `models/stt/cohere-transcribe-03-2026/coreml/` — no changes to VAD, Qwen3, or other model directories.

### Exports (`exports/`)
- `export-encoder.py` — 48-layer Conformer encoder, fixed mel input `[1, 128, 3500]`, output `[1, 438, 1024]`. Uses the correct `length` masking parameter so padded frames are ignored internally.
- `export-decoder-cache-external.py` — 8-layer decoder with Parakeet-style **external KV cache** (108-token window). This is the shipped decoder path; the stateful-decoder variant was dropped after the cache-external version landed.

### Tools (`tools/`)
- `cohere_features_v2.py` — canonical numpy port of `FilterbankFeatures` (see host-side fixes below).
- `quantize_to_int8.py` — W8A16 quantization for the encoder.
- `compile_encoder_to_mlmodelc.py` — `.mlpackage` → `.mlmodelc` compile step with ANE targeting.
- `download-fleurs-for-swift.py` — FLEURS slice downloader used by the Swift benchmark.

### Host-side pipeline fixes (the important part)
The shipped `cohere_mel_spectrogram.py` did not match `processing_cohere_asr.py::FilterbankFeatures` on any parameter that matters: wrong `n_fft`, wrong window, wrong mel normalization, wrong log base, and **no per-feature CMVN at all**. Without CMVN, every utterance's features drift by tens of dB per bin, the encoder is fed out-of-distribution data, and the decoder emits whatever language cluster happened to be nearest — Arabic for French, Polish for Chinese, etc.

Four host-only fixes (no retraining, same weights) make the failures disappear:

1. **`tools/cohere_features_v2.py`** — faithful numpy port of `FilterbankFeatures`: `n_fft=512`, Hann(400) zero-padded to 512, preemph=0.97, Slaney mel, natural log + `2^-24` guard, per-feature CMVN ddof=1 ε=1e-5, `mag_power=2.0`. Verified against `AutoFeatureExtractor.from_pretrained(..., trust_remote_code=True)` on real samples × 4 languages via `tests/test-feature-parity.py` — residual is within HF's own dither variance.
2. **Cross-attention masking** — the encoder always emits 438 frames but only `ceil(feature_length * 438/3500)` correspond to real audio. Padded frames are masked with `-1e4`.
3. **Repetition penalty + no-repeat-ngram** in greedy decode (defaults 1.1 and 3).
4. **SentencePiece byte-fallback detok** — CJK characters emit as `<0xHH>` triples; `tokens_to_text` buffers and flushes via `bytes(...).decode(\"utf-8\", errors=\"replace\")`.

### FLEURS impact (3 samples × 4 languages, same CoreML model files)

| Language | Metric | OLD | NEW | Δ |
|---|---|---|---|---|
| en_us | WER | 55.3% | 10.6% | **−44.6 pp** |
| es_419 | WER | 11.3% | 4.9% | **−6.4 pp** |
| fr_fr | WER | 92.1% | 16.8% | **−75.2 pp** |
| cmn_hans_cn | CER | 261.7% | 14.1% | **−247.6 pp** |

## Q8 findings

The shipped q8 decoder has a separate, orthogonal failure mode from the host-side bugs: **over-generation**. It produces a correct transcript, then keeps going past the real EOS. Instrumented logging shows EOS is consistently **rank 1 or 2** with a ~2 logit gap to the winner — textbook weight-only INT8 error on the final classifier.

Re-quantization experiments (kept decoder lm_head at FP16 via tied-const handling, per-tensor variants, threshold skipping) never matched FP16; the q8 quality loss is distributed across many layers, not localized to lm_head. A proper fix would need calibration-aware quantization with end-of-utterance frames, or mixed-precision keeping attention output projections at FP16 — both outside the `coremltools.optimize.coreml` op-level API.

**Recommendation adopted downstream:** ship **INT8 encoder (W8A16) + FP16 cache-external decoder** as the hybrid default. The encoder INT8 is lossless (±0.5% WER) and the FP16 decoder sidesteps the EOS regression entirely. See `docs/HOST_SIDE_FIXES.md` and `docs/Q8_EOS_BIAS.md` for full detail.

## Downstream validation (FluidAudio#487)

Swift integration with the INT8 encoder + FP16 cache-external decoder is live in [FluidAudio#487](https://github.com/FluidInference/FluidAudio/pull/487). Benchmarks using the exports and feature extractor from this PR:

### LibriSpeech test-clean (full split, Apple M2 2022)

| Samples | WER | CER | RTFx (per-file mean) | RTFx (total) |
|---:|---:|---:|---:|---:|
| 2,620 | **1.77%** | **0.60%** | 2.04× | 1.72× |

### FLEURS (full splits, 14 languages, M4 Pro)

| Code | Lang | Samples | WER | CER | RTFx |
|---|---|---:|---:|---:|---:|
| en_us | English | 647 | 5.63% | 3.19% | 2.49× |
| fr_fr | French | 676 | 6.22% | 3.11% | 2.21× |
| de_de | German | 862 | 5.84% | 2.83% | 1.98× |
| es_419 | Spanish (LatAm) | 908 | 4.53% | 2.40% | 1.34× |
| it_it | Italian | 865 | **4.03%** | 2.04% | **3.15×** |
| pt_br | Portuguese (BR) | 919 | 6.44% | 3.38% | 2.79× |
| nl_nl | Dutch | 364 | 8.07% | 4.14% | 2.04× |
| pl_pl | Polish | 758 | 7.49% | 3.23% | 1.98× |
| el_gr | Greek | 650 | 11.50% | 5.45% | 2.00× |
| ar_eg | Arabic (EG) | 428 | 18.46% | 6.71% | 2.06× |
| ja_jp | Japanese | 650 | 60.13%† | **6.25%** | 2.23× |
| cmn_hans_cn | Mandarin (Simp) | 945 | 98.52%† | **12.01%** | 1.85× |
| ko_kr | Korean | 382 | 16.39% | 6.67% | 1.84× |
| vi_vn | Vietnamese | 857 | 9.55% | 6.87% | 1.55× |

†ja/zh written without word boundaries → WER is tokenization artifact; CER is the real accuracy metric for those languages.

Comparison vs. Cohere's technical-report Figure 4 (FLEURS+CV17+MLS+Wenet avg, FP16 PyTorch): ours lands ~1–3 pp higher on most languages, consistent with (a) FLEURS-only being harder than a 4-corpus average and (b) the INT8 encoder. Japanese CER is actually **below** Cohere's reported number.

## Docs landed in this PR

- `coreml/README.md` — end-to-end CoreML conversion walkthrough.
- `docs/HOST_SIDE_FIXES.md` — the four host-side fixes and FLEURS A/B tables.
- `docs/Q8_EOS_BIAS.md` — EOS-bias analysis and re-quantization experiments.
- `docs/CACHE_EXTERNAL_ANALYSIS.md`, `CACHE_EXTERNAL_DELIVERED.md`, `CACHE_INVESTIGATION_SUMMARY.md` — cache-external decoder path rationale and delivery notes.
- `docs/COHERE_ARCHITECTURE_ANALYSIS.md` — upstream model walkthrough and CoreML mapping.

## Test plan

- [x] Encoder export with correct `length` masking
- [x] Cache-external decoder export (external KV cache, 108-token window)
- [x] INT8 W8A16 encoder quantization
- [x] `cohere_features_v2.py` parity vs HF `AutoFeatureExtractor` (`tests/test-feature-parity.py`)
- [x] FLEURS 4-language A/B (broken vs fixed host pipeline)
- [x] HuggingFace uploads (f16 and q8 hybrid)
- [x] Swift integration landed in [FluidAudio#487](https://github.com/FluidInference/FluidAudio/pull/487)
- [x] LibriSpeech test-clean full benchmark (2,620 samples, WER 1.77%)
- [x] Full 14-language FLEURS benchmark

## Known limitations

- Fully-q8 decoder is unusable without a runtime EOS bias or re-quantization with EOU calibration; ship INT8-encoder + FP16-decoder hybrid instead.
- INT4 encoder rejected (293% avg WER).
- Swift CoreML pipeline is single-chunk (≤35 s per call); >35 s audio requires the upstream 5-s-overlap sliding window, not yet ported.

### CODE REVIEW

File: models/stt/cohere-transcribe-03-2026/cohere-pytorch/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,37 @@
+*.7z filter=lfs diff=lfs merge=lfs -text
```
Comment (BrandonWeng) [inline models/stt/cohere-transcribe-03-2026/cohere-pytorch/.gitattributes]: no lfs pls. do not commit here
